pages/DSNP/Graph.md

```diff
@@ -2,29 +2,124 @@
+
+PRId objects MUST comply with the following [Avro](https://avro.apache.org) schema:
+
+```
+{
+    "namespace": "org.dsnp.userdata",
+    "name": "PRId",
+    "type": "fixed",
+    "size": 16,
+    "doc": "Pseudonymous Relationship Identifier"
+}
+```
+
+More generally, a PRId is a means of allowing for private interactions between two DSNP users in a specific context, based on cryptographic primitives that enable two users to independently generate a shared secret.
+For each pair of users Alice and Bob, and a specified PRId [context](#contexts), a pair of PRIds can be generated from Alice and Bob's shared secret, with one representing a simplex channel from Alice to Bob, and the other representing the corresponding simplex channel from Bob to Alice.
+
+PRIds have the following properties:
+* If Alice has generated a PRId from Alice to Bob, only Bob can generate the same PRId. This allows Bob to check if he is listed in Alice's published connections.
+* Bob can prove to a third party that a given PRId in Alice's published connections belongs to him, without the third party being able to know anything about any of Alice or Bob's relationships with any other user.
+
+#### Algorithm
+
+In the following section, the Alice to Bob identifier for context C is called PRId<sub>A→B,C</sub>, and the corresponding Bob to Alice identifier is called PRId<sub>B→A,C</sub>.
+
+A PRId is derived from Alice and Bob's `keyAgreement` key pairs, using a key exchange protocol as follows. To illustrate the cryptographic operations required, the relevant functions from [libsodium](https://libsodium.org) are noted. Sodium is a stable, fast, free, and cross-platform cryptography library, and supports all encryption algorithms used in the DSNP specification out of the box.
+
+1. Both Alice and Bob generate an asymmetric key pair, and each publishes a Public Key Announcement with their generated public key and a `keyType` value of `keyAgreement`.
+1. When Alice wants to interact with Bob, she looks up Bob's public key and performs a key exchange operation using her private key and Bob's public key, generating a shared secret `k` as in the libsodium function <code>crypto_box_beforenm(k, Bob<sub>public</sub>, Alice<sub>private</sub>)</code>.
+1. Alice derives a context-specific subkey <code>k<sub>PRId</sub></code>, using the shared secret `k` as the master key, as in the libsodium function `crypto_kdf_derive_from_key`, using `0` (zero) as the key identifier (see the [PRId Context](#contexts) table below) and the ASCII encoding of `"DSNPPRId"` as the 8-byte human-readable identifier.
+1. Alice concatenates her DSNP User Id and Bob's DSNP User Id to form a 16-byte message.
+1. Alice encrypts the message using the PRId key <code>k<sub>PRId</sub></code> and a nonce of all zeros, as in the libsodium function `crypto_secret_box_easy`.
```

Comment:
Yes. And good point, the same nonce shouldn't be reused with the same key. I need to change it so that the derived key is specific to the relationship target... I forgot about Bob using the same key/nonce. If the derived keys are different, a fixed nonce is OK because the key is only ever used for this one encryption.